migrate_pim/upgrade_from_32_to_40.rst

```diff
@@ -229,3 +229,139 @@ From that point, you will have to migrate your bundle one by one.
+                    $attribute->code(),
+                    static::class,
+                    $data
+                );
+            }
+
+            if (!isset($data['min'])) {
+                throw InvalidPropertyTypeException::arrayKeyExpected(
+                    $attribute->code(),
+                    'min',
+                    static::class,
+                    $data
+                );
+            }
+
+            if (null === $data['min'] && null === $data['max']) {
+                throw InvalidPropertyException::valueNotEmptyExpected(
+                    $attribute->code(),
+                    static::class
+                );
+            }
+
+            if (!isset($data['max'])) {
+                throw InvalidPropertyTypeException::arrayKeyExpected(
+                    $attribute->code(),
+                    'max',
+                    static::class,
+                    $data
+                );
+            }
+
+            return $this->createWithoutCheckingData($attribute, $channelCode, $localeCode, $data);
+        }
+
+        public function createWithoutCheckingData(Attribute $attribute, ?string $channelCode, ?string $localeCode, $data): ValueInterface
+        {
+            if ($attribute->isLocalizableAndScopable()) {
+                return RangeValue::scopableLocalizableValue($attribute->code(), $data, $channelCode, $localeCode);
+            }
+            if ($attribute->isScopable()) {
+                return RangeValue::scopableValue($attribute->code(), $data, $channelCode);
+            }
+            if ($attribute->isLocalizable()) {
+                return RangeValue::localizableValue($attribute->code(), $data, $localeCode);
+            }
+
+            return RangeValue::value($attribute->code(), $data);
+        }
+
+        public function supportedAttributeType(): string
+        {
+            return RangeType::RANGE;
+        }
+    }
+
+You have to declare it in a new registry as well in the dependency injection configuration, named `akeneo.pim.enrichment.factory.product_value`.
+
+.. note::
+    ```yaml
+        acme.range.factory.value.range:
+            class: 'Acme\RangeBundle\Product\Factory\Value\RangeValueFactory'
+            tags: ['akeneo.pim.enrichment.factory.product_value']
+    ```
+
+In the first method, `createByCheckingData`, the data type should be checked. For example, it checks that an expected scalar is indeed a scalar. This is done to guarantee that data manipulated in the domain layer are corrects. This method is useful when data are coming from the outside world (product save in UI, API, import, etc). This validation is costly as soon as you have several hundreds values per product.
+
+That's why we implemented a second method to avoid to do these checks. It should be used when a value collection is created from values coming from the database, as the data is already validated and consistent in Mysql. It avoid to pay the performance penalty of checking types.
```

Comment:
```suggestion
That's why we implemented a second method to avoid to do these checks. It should be used when a value collection is created from values coming from the database, as the data is already validated and consistent in Mysql. It avoids to pay the performance penalty for checking types.
```